File: Sorting/Sorting/Algorithms/MergeSort.cs
namespace Sorting.Algorithms
{
	using System.Collections.Generic;
	using System.Linq;

	public class MergeSort: SortBase
	{
		private int[] _copy;

		public override IEnumerable<int> Execute(ICollection<int> unsortedList)
		{
			base.Execute(unsortedList);
			_copy = Copy(unsortedList).ToArray();
			AddToIntermediateSorts(_copy);
			
			TopDownMergeSort(_copy, new int[_copy.Length], _copy.Length);

			StopWatch();
			return _copy;
		}
		
		private void TopDownMergeSort(int[] a, int[] b, int n)
		{
			CopyArray(a, 0, n, b);
			TopDownSplitMerge(b, 0, n, a);
		}

		private void TopDownSplitMerge(int[] b, int iBegin, int iEnd, int[] a)
		{
			if (iEnd - iBegin < 2)
			{
				return;
			}

			var iMiddle = (iEnd + iBegin) / 2;

			TopDownSplitMerge(a, iBegin, iMiddle, b);
			TopDownSplitMerge(a, iMiddle, iEnd, b);
			
			TopDownMerge(b, iBegin, iMiddle, iEnd, a);
		}

		private void TopDownMerge(int[] a, int iBegin, int iMiddle, int iEnd, int[] b)
		{
			var i = iBegin;
			var j = iMiddle;
			
			for (var k = iBegin; k < iEnd; k++)
			{
				if (i < iMiddle && (j >= iEnd || a[i] <= a[j]))
				{
					b[k] = a[i];
					i = i + 1;
				}
				else
				{
					b[k] = a[j];
					j = j + 1;
				}
			}

			AddToIntermediateSorts(_copy);
		}

		private void CopyArray(int[] a, int iBegin, int iEnd, int[] b)
		{
			for (var k = iBegin; k < iEnd; k++)
			{
				b[k] = a[k];
			}
		}
	}
}

File: Sorting/Sorting.WebPure/scripts/Algorithms/MergeSort.ts
namespace Algorithms {
	export class MergeSort extends SortBase {
		private arrayCopy: number[];

		public execute(unsortedList: number[]) {
			super.execute(unsortedList);
			this.arrayCopy = super.copy(unsortedList);
			super.addToIntermediateSorts(this.arrayCopy);
			
			this.topDownMergeSort(this.arrayCopy, [], this.arrayCopy.length);

			//StopWatch();
			return this.arrayCopy;
		}

		private topDownMergeSort(a: number[], b: number[], n: number): void {
			this.copyArray(a, 0, n, b);
			this.topDownSplitMerge(b, 0, n, a);
		}

		private topDownSplitMerge(b: number[], iBegin: number, iEnd: number, a: number[]): void {
			if (iEnd - iBegin < 2) {
				return;
			}

			const iMiddle = Math.floor ((iEnd + iBegin) / 2);
			this.topDownSplitMerge(a, iBegin, iMiddle, b);
			this.topDownSplitMerge(a, iMiddle, iEnd, b);

			this.topDownMerge(b, iBegin, iMiddle, iEnd, a);
		}

		private topDownMerge(a: number[], iBegin: number, iMiddle: number, iEnd: number, b: number[]): void {
			let i = iBegin;
			let j = iMiddle;

			for (let k = iBegin; k < iEnd; k++) {
				if (i < iMiddle && (j >= iEnd || a[i] <= a[j])) {
					b[k] = a[i];
					i++;
				}
				else {
					b[k] = a[j];
					j++;
				}
			}

			super.addToIntermediateSorts(this.arrayCopy);
		}

		private copyArray(a: number[], iBegin: number, iEnd: number, b: number[]): void {
			for (let k = iBegin; k < iEnd; k++) {
				b[k] = a[k];
			}
		}
	}
}